### PR TITLE
CI: unpin numpy for CI / Checks github action

### PR DESCRIPTION
**Pull request by simonjayhawkins**
xref #39513

### CODE REVIEW

**Comment by WillAyd** [comment]
Cool thanks for taking a look. Are you expecting this new environment to be part of CI or to just leave here for visibility as we clean up the existing failures?

**Comment by simonjayhawkins** [comment]
> Cool thanks for taking a look. Are you expecting this new environment to be part of CI or to just leave here for visibility as we clean up the existing failures?

open to suggestions.

vendoring stub or adding additional steps to current check (update environment and run mypy again) are other options.

**Comment by jbrockmendel** [comment]
@simonjayhawkins is getting this passing viable?  a bunch of the complaints are about e.g. `np.random.randn`, not clear how to resolve

**Comment by simonjayhawkins** [comment]
> a bunch of the complaints are about e.g. `np.random.randn`, not clear how to resolve

probably need to submit some fixes upstream

> is getting this passing viable?

imo not with the current review process, even though I normally put the errors in the OP for typing fixes

I'm now going to add the errors directly to code here, so it that it may be easier to identify the problems (and the low hanging fruit)

I may open another PR that sits on top of this with the fixes so that it is easier to see the reason for the code changes from the diff

File: pandas/plotting/_matplotlib/converter.py [hunk 1]
```diff
@@ -891,16 +885,11 @@ class TimeSeries_DateLocator(Locator):
...
-        Whether the locator should work in dynamic mode.
-    base : {int}, optional
-    quarter : {int}, optional
-    month : {int}, optional
-    day : {int}, optional
+    freq : {var} Valid frequency specifier. minor_locator : {False, True},
+        optional Whether the locator is for minor ticks (True) or not.
+        dynamic_mode : {True, False}, optional Whether the locator should work
+        in dynamic mode. base : {int}, optional quarter : {int}, optional month
+        : {int}, optional day : {int}, optional
```
Comment (jbrockmendel) [inline pandas/plotting/_matplotlib/converter.py:892]: is this edit intentional?  looks like a few other docstrings in this module got mangled

Comment (simonjayhawkins) [reply pandas/plotting/_matplotlib/converter.py:892]: Thanks @jbrockmendel . I'd noticed that a few docstrings got autoformatted. This PR is just for tracking at the moment, but will sort that out at some point.

**Comment by jbrockmendel** [comment]
Looks like a lot of unused "type:ignore" complaints, but otherwise mypy is happy?

**Comment by simonjayhawkins** [comment]
> > a bunch of the complaints are about e.g. `np.random.randn`, not clear how to resolve
> 
> probably need to submit some fixes upstream

has now been fixed upstream

File: pandas/core/algorithms.py [hunk 2]
```diff
@@ -2214,7 +2214,9 @@ def _sort_mixed(values):
     return np.concatenate([nums, np.asarray(strs, dtype=object)])
 
 
-def _sort_tuples(values: np.ndarray[tuple]):
+# pandas\core\algorithms.py:2217: error: "ndarray" expects no type arguments,
+# but 1 given  [type-arg]
+def _sort_tuples(values: np.ndarray[tuple]):  # type: ignore[type-arg]
```
Comment (jbrockmendel) [inline pandas/core/algorithms.py:2219]: IIUC np.ndarray takes annotations of the form `ndarray[shape, dtype]`, not sure that `tuple` make sense in this context

Comment (simonjayhawkins) [reply pandas/core/algorithms.py:2219]: ndarray doesn't take type arguments (is not a generic type). so this just needs to be changed to just `np.ndarray` on master.

https://github.com/numpy/numpy/blob/a74789081fd9753c5e5637c404523b8f47b55dcc/numpy/__init__.pyi#L1478

**Comment by simonjayhawkins** [comment]
> Looks like a lot of unused "type:ignore" complaints, but otherwise mypy is happy?

the ignores are with numpy stable. Found 402 errors in 81 files. That's the number of ignores to get mypy to pass with numpy nightly. (added as separate ci job)

we should have a discussion at some point how we plan to 'switch' to using numpy 1.20 for type checking once released.

**Comment by simonjayhawkins** [comment]
> > Looks like a lot of unused "type:ignore" complaints, but otherwise mypy is happy?
> 
> the ignores are with numpy stable. Found 402 errors in 81 files. That's the number of ignores to get mypy to pass with numpy nightly. (added as separate ci job)
> 
> we should have a discussion at some point how we plan to 'switch' to using numpy 1.20 for type checking once released.

xref #37740 reduces to Found 312 errors in 75 files if we can get buy-in for change AnyArrayLike and ArrayLike to Union instead of TypeVars.  https://github.com/pandas-dev/pandas/pull/37224#discussion_r512859195

**Comment by jbrockmendel** [comment]
>  if we can get buy-in for change AnyArrayLike and ArrayLike to Union instead of TypeVars

I'm on board.

**Comment by simonjayhawkins** [comment]
#37771 fixed one of the mypy errors here. There may be more errors hidden from mypy on master because numpy types currently resolve to `Any` that are not false positives once we can remove much of the noise here.

**Comment by simonjayhawkins** [comment]
I've added the 1.2.2 milestone. see https://github.com/pandas-dev/pandas/issues/39513#issuecomment-771006563

This PR is moreless just comments, so should be no harm backporting

an alternative maybe to modify the ci script on 1.2.x to use a different env from environment.yml  or to not follow imports when mypy checking

**Comment by simonjayhawkins** [comment]
This gets mypy to green. 

benchmarks failing, see  https://github.com/pandas-dev/pandas/pull/38379#discussion_r567753143

need to deal with that (as a pre-cursor) before unpining numpy

**Comment by jreback** [comment]
if you would merge master

**Comment by simonjayhawkins** [comment]
moving to 1.2.3

File: pandas/compat/numpy/__init__.py [hunk 3]
```diff
@@ -6,7 +6,9 @@
 import numpy as np
 
 # numpy versioning
-_np_version = np.__version__
+# pandas\compat\numpy\__init__.py:9: error: Module has no attribute
+# "__version__"; maybe "version"?  [attr-defined]
+_np_version = np.__version__  # type: ignore[attr-defined]
```
Comment (simonjayhawkins) [inline pandas/compat/numpy/__init__.py:11]: fixed in numpy 1.20.1

File: pandas/compat/numpy/function.py [hunk 4]
```diff
@@ -18,7 +18,9 @@
 from distutils.version import LooseVersion
 from typing import Any, Dict, Optional, Union
 
-from numpy import __version__, ndarray
+# pandas\compat\numpy\function.py:23: error: Module 'numpy' has no attribute
+# '__version__'; maybe "version"?  [attr-defined]
+from numpy import __version__, ndarray  # type: ignore[attr-defined]
```
Comment (simonjayhawkins) [inline pandas/compat/numpy/function.py:23]: same

File: pandas/core/frame.py [hunk 5]
```diff
@@ -2097,13 +2208,18 @@ def to_records(
...
+                # "Union[type, dtype, str]"; expected "dtype"
+                formats.append(dtype_mapping)  # type: ignore[arg-type]
             else:
                 element = "row" if i < index_len else "column"
                 msg = f"Invalid dtype {dtype_mapping} specified for {element} {name}"
                 raise ValueError(msg)
 
-        return np.rec.fromarrays(arrays, dtype={"names": names, "formats": formats})
+        # error: Module has no attribute "fromarrays"
+        return np.rec.fromarrays(  # type: ignore[attr-defined]
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:2220]: fixed in numpy 1.20.1

File: pandas/core/algorithms.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/algorithms.py:2318]: opened #39792

**Comment by simonjayhawkins** [comment]
urgh. IsInLongSeriesValuesDominate is also failing. didn't notice that before.

**Comment by simonjayhawkins** [comment]
green. 

have added the benchmark skip here https://github.com/pandas-dev/pandas/pull/36092/commits/4068a38c858bc4be81efb4a66e6ec7218594c847. this could be a pre-cursor. ( addresses @jorisvandenbossche follow-up comment https://github.com/pandas-dev/pandas/pull/39795#discussion_r576792892)

**Comment by jreback** [comment]
@simonjayhawkins we want to backport this?

**Comment by simonjayhawkins** [comment]
I don't think so, but we should remove the pins from environment.yml on 1.2.x at the same time. will need to change the ci workflow file to use the old numpy or not follow imports for typing. 

This can be done after or before this.

**Comment by simonjayhawkins** [comment]
so the issue this is addressing should remain milestoned 1.2.x and this issue does not close it.

**Comment by simonjayhawkins** [comment]
> I don't think so, but we should remove the pins from environment.yml on 1.2.x at the same time. will need to change the ci workflow file to use the old numpy or not follow imports for typing.
> 
> This can be done after or before this.

#39998 and #39994

**Comment by jreback** [comment]
this lgtm @simonjayhawkins rebase once more just to be sure and let's merge!

**Comment by simonjayhawkins** [comment]
ok. will merge shortly if green (hopefully will be, one 1 commit to master since last rebase)

There will likely be some ci errors with merging subsequent PRs that are not rebased to this one. I _think_ that we could take the hit and follow-up with ci fixes rather than asking too many PR authors to rebase.

**Comment by jreback** [comment]
thanks @simonjayhawkins very nice!

**Comment by jreback** [comment]
> ok. will merge shortly if green (hopefully will be, one 1 commit to master since last rebase)
> 
> There will likely be some ci errors with merging subsequent PRs that are not rebased to this one. I _think_ that we could take the hit and follow-up with ci fixes rather than asking too many PR authors to rebase.

yep we'll see and that's fine too
